src/link/script.y

```diff
@@ -0,0 +1,479 @@
+                                               context.path.c_str(), context.lineNo, __VA_ARGS__)
+// MSVC doesn't support __VA_OPT__ yet.
+#define scriptErrorSimple(context, str) ::error(NULL, 0, "%s(%" PRIu32 "): " str, \
+                                               context.path.c_str(), context.lineNo)
+
+// Lexer.
+
+struct LexerStackEntry {
+	std::filebuf file;
+	std::string path;
+	uint32_t lineNo;
+
+	using int_type = decltype(file)::int_type;
+	static constexpr int_type eof = decltype(file)::traits_type::eof();
+
+	explicit LexerStackEntry(std::string &&path_) : file(), path(path_), lineNo(1) {}
+};
+static std::vector<LexerStackEntry> lexerStack;
+
+void yy::parser::error(std::string const &msg) {
+	auto const &script = lexerStack.back();
+	scriptError(script, "%s", msg.c_str());
+}
+
+static void includeFile(std::string &&path) {
+	// `emplace_back` can invalidate references to the stack's elements!
+	// This is why `newContext` must be gotten before `prevContext`.
+	auto &newContext = lexerStack.emplace_back(std::move(path));
+	auto &prevContext = lexerStack[lexerStack.size() - 2];
+
+	if (!newContext.file.open(newContext.path, std::ios_base::in)) {
+		// The order is important: report the error, increment the line number, modify the stack!
+		scriptError(prevContext, "Could not open included linker script \"%s\"",
+		            newContext.path.c_str());
+		++prevContext.lineNo;
+		lexerStack.pop_back();
+	} else {
+		// The lexer will use the new entry to lex the next token.
+		++prevContext.lineNo;
+	}
+}
+
+static void incLineNo(void) {
+	++lexerStack.back().lineNo;
+}
+
+static bool isWhiteSpace(LexerStackEntry::int_type c) {
+	return c == ' ' || c == '\t';
+}
+
+static bool isNewline(LexerStackEntry::int_type c) {
+	return c == '\r' || c == '\n';
+}
+
+static bool isIdentChar(LexerStackEntry::int_type c) {
+	return (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9');
+}
+
+static bool isHexDigit(LexerStackEntry::int_type c) {
+	return (c >= '0' && c <= '9') || (c >= 'A' && c <= 'F') || (c >= 'a' && c <= 'f');
+}
+
+static uint8_t parseHexDigit(LexerStackEntry::int_type c) {
+	if (c >= '0' && c <= '9') {
+		return c - '0';
+	} else if (c >= 'A' && c <= 'F') {
+		return c - 'A' + 10;
+	} else if (c >= 'a' && c <= 'f') {
+		return c - 'a' + 10;
+	} else {
+		unreachable_();
+	}
+}
+
+yy::parser::symbol_type yylex(void) {
+try_again: // Can't use a `do {} while(0)` loop, otherwise compilers (wrongly) think it can end.
+	auto &context = lexerStack.back();
+	auto c = context.file.sbumpc();
+
+	// First, skip leading whitespace.
+	while (isWhiteSpace(c)) {
+		c = context.file.sbumpc();
+	}
+	// Then, skip a comment if applicable.
+	if (c == ';') {
+		while (!isNewline(c)) {
+			c = context.file.sbumpc();
+		}
+	}
+
+	// Alright, what token should we return?
+	if (c == LexerStackEntry::eof) {
+		// Basically yywrap().
+		lexerStack.pop_back();
+		if (!lexerStack.empty()) {
+			goto try_again;
```

Comment:
I thought a tail-call `return yylex();` would be optimized correctly. We do that in `peek()` in asm/lexer.cpp.

Reply:
At least not in debug mode, since there stack traces are relied upon. I think `goto` is more appropriate, if only to express intent.

Reply:
It's tolerable, then. :P